plugins/SegmentEditor/API.php

```diff
@@ -238,29 +229,42 @@ private function getModel()
      * @param int $idSegment The ID of the stored segment to modify.
      * @param string $name The new name of the segment.
      * @param string $definition The new definition of the segment.
-     * @param bool $idSite If supplied, associates the stored segment with as single site.
+     * @param int|string|null $idSite If supplied, associates the stored segment with as single site.
      * @param bool $autoArchive Whether to automatically archive data with the segment or not.
      * @param bool $enabledAllUsers Whether the stored segment is viewable by all users or just the one that created it.
-     *
-     * @return bool
      */
-    public function update($idSegment, $name, $definition, $idSite = false, $autoArchive = false, $enabledAllUsers = false)
-    {
+    public function update(
+        int $idSegment,
+        string $name,
+        string $definition,
+        $idSite = null,
+        bool $autoArchive = false,
+        bool $enabledAllUsers = false
+    ): void {
         $segment = $this->getSegmentOrFail($idSegment);
         $this->checkUserCanEditOrDeleteSegment($segment);
 
         $idSite = $this->checkIdSite($idSite);
+        $name = Common::sanitizeInputValue($name);
         $this->checkSegmentName($name);
-        $definition      = $this->checkSegmentValue($definition, $idSite);
-        $enabledAllUsers = $this->checkEnabledAllUsers($enabledAllUsers);
+        $definition = Common::sanitizeInputValue($definition);
+        $definition = $this->checkSegmentValue($definition, $idSite);
+
+        // only check param if value is changed
+        // this ensure that a segment from a user with lower permission can still be changed by them
+        // if a superuser updated the segment to be available for all users
+        if ((int) $segment['enable_all_users'] !== (int) $enabledAllUsers) {
+            $enabledAllUsers = $this->checkEnabledAllUsers($enabledAllUsers);
```

Comment:
This change currently has an interesting effect if a user wants to claim his segment back from the general public:

A regular user is always allowed to set `enable_all_users = 0`. So if someone uses a direct API request, or has a browser state loaded where that flag is still set to "single user segment", any update will reset the super user sharing.

I guess an exception saying "segment shared to all users, sharing can only be changed by super user" can't hurt, unless we accept a user to revoke the shared state "if he knows how to do that".

_Already updated by showing an error message, keeping this just for documentation._